pkg/util/api.go

```diff
@@ -75,17 +77,72 @@ func RequestBodyToBytes(request *http.Request) ([]byte, error) {
 		return nil, err
 	}
 
-	if strings.Contains(contentType, "application/json") && !json.Valid(body) {
-		return nil, &HTTPError{
-			ErrorType:   "BadRequest",
-			Description: "request body is not valid JSON",
-			StatusCode:  http.StatusBadRequest,
+	if strings.Contains(contentType, "application/json") {
+		if err := validJson(body); err != nil {
+			return nil, &HTTPError{
+				ErrorType:   "BadRequest",
+				Description: fmt.Sprintf("Request body is not valid: %s", err),
+				StatusCode:  http.StatusBadRequest,
+			}
 		}
 	}
 
 	return body, nil
 }
 
+func validJson(data []byte) error {
+	if !json.Valid(data) {
+		return fmt.Errorf("invalid json")
+	}
+	return checkDuplicateKeys(json.NewDecoder(bytes.NewReader(data)), nil)
+}
+
+func checkDuplicateKeys(d *json.Decoder, path []string) error {
+	t, err := d.Token()
```

Comment:
you can also unmarshal and marshal i think - it would be probably slower but easier to maintain - not sure whether it makes sense

Reply:
Unmarshaling and Marshaling gets the last provided value for objects (if the value in which you unmarshal in is struct) and merge the values for duplicate keys if unmarshal in map which makes it impossible to know which is the right value, we can't know what json is provided to rely on this. It can be manipulated in many ways, this is why most of other language JSON parsers does not allow duplicate keys and I decided to go in this direction.